tf_encrypted/protocol/securenn.py

```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
         """
         return self.dispatch('equal_zero', x, container=_thismodule, out_dtype=out_dtype)
 
-    def share_convert(self, x):
+    def share_convert(self, a):
         """
         Convert which ring `x` belongs to.  This protocol is not implemented yet.
 
         Some operations in secureNN only work in an odd ring.  This function
         is used to convert from one ring to another.
 
         :param PondTensor x: The tensor to convert.
-        :raises: NotImplementedError
         """
-        raise NotImplementedError
+
+        L = self.tensor_factory.modulus
+
+        if L > 2**64:
+            raise Exception('SecureNN share convert only support moduli of less or equal to 2 ** 64.')
+
+        with tf.device(self.server_0.device_name):
+            # Common randomness
+            npp = _generate_random_bits(self, a.shape)
+            r = self.tensor_factory.sample_uniform(a.shape)
+            r_0, r_1 = self._share(r)
+            r = PondPublicTensor(self, r_0 + r_1, r_0 + r_1, is_scaled=False)
+            alpha = self.compute_wrap(r_0, r_1, L)
+
+        with tf.device(self.server_0.device_name):
+            # line 2
+            a_hat_0 = a.share0 + r_0
+            beta_0 = self.compute_wrap(a.share0, r_0, L)
+
+        with tf.device(self.server_1.device_name):
+            # line 2
+            a_hat_1 = a.share1 + r_1
+            beta_1 = self.compute_wrap(a.share1, r_1, L)
+
+        # line 3
+        with tf.device(self.crypto_producer.device_name):
+            # line 4
+            a_hat = PondPrivateTensor(self, a_hat_0, a_hat_1, is_scaled=False)
+            x = a_hat.reveal()
+            gamma = self.compute_wrap(a_hat.share0, a_hat.share1, L)
+
+            # line 5
+            x_bits = x.value_on_0.to_bits(self.prime_factory)
+            x_bits = self._share_and_wrap(x_bits, is_scaled=False)
+            # need shares of gamma in L-1
+            gamma_odd = gamma.to_odd_modulus(self.implicit_factory)
+            gamma = self._share_and_wrap(gamma_odd, is_scaled=False)
+
+        # line 6
+        np = _private_compare(self, x_bits, r - 1, npp)
+
+        # line 7 (convert np to L-1)
+        with tf.device(self.crypto_producer.device_name):
+            np_0, np_1 = np.reveal().unwrapped
+            np_0 = np_0.to_odd_modulus(self.implicit_factory)
+
+            np_odd = self._share_and_wrap(np_0, is_scaled=False)
+
+        # line 9
+        # n = np_odd + npp - np_odd * npp * 2
+
+        # line 10, j = 0
+        with tf.device(self.server_0.device_name):
+            n_0 = np_odd.share0 + npp.value_on_0 - (np_odd.share0 * npp.value_on_0 * 2)
+            a_0 = a.share0.to_odd_modulus(self.implicit_factory)
+            theta_0 = beta_0.to_odd_modulus(self.implicit_factory) + (((-alpha).to_odd_modulus(self.implicit_factory) - 1) * 1) + gamma.share0 + n_0
```

Comment:
I think you should be able to compute `theta` without using `tf.device`

Reply:
I think we'd need to write casting for pond tensors which I can do but might be a little out of scope here.